server/src/main/java/io/crate/execution/dml/ArrayIndexer.java

```diff
@@ -26,22 +26,48 @@
 import java.util.function.Consumer;
 import java.util.function.Function;
 
-import org.jetbrains.annotations.NotNull;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.IntField;
 import org.jetbrains.annotations.Nullable;
 
 import io.crate.metadata.ColumnIdent;
 import io.crate.metadata.Reference;
 
 public class ArrayIndexer<T> implements ValueIndexer<List<T>> {
 
+    public static <T> ArrayIndexer<T> beforeV590(ValueIndexer<T> innerIndexer, Reference reference) {
```

Comment:
I think a less invasive change would be to add the table created version to IndexDocumentBuilder, and then check the version in `ArrayIndexer.indexValue()` and add the length field if it's 5.9 or later.  Then we don't need the changes to all the DataTypes and Indexers, we only have a single ArrayIndexer implementation, etc.

The benchmark results are very impressive!

Reply:
I'd be surprised if an extra comparison really made a difference, and I think the smaller API surface (fewer parameters passed to StorageSupport, etc) is worth it.